File: src/serial/CondFormats/PixelCPEFast.cc
#include <iostream>
#include <fstream>

#include "Geometry/phase1PixelTopology.h"
#include "CondFormats/PixelCPEFast.h"

// Services
// this is needed to get errors from templates

namespace {
  constexpr float micronsToCm = 1.0e-4;
}

//-----------------------------------------------------------------------------
//!  The constructor.
//-----------------------------------------------------------------------------
PixelCPEFast::PixelCPEFast(std::string const &path) {
  {
    std::ifstream in(path, std::ios::binary);
    in.exceptions(std::ifstream::badbit | std::ifstream::failbit | std::ifstream::eofbit);
    in.read(reinterpret_cast<char *>(&m_commonParamsGPU), sizeof(pixelCPEforGPU::CommonParams));
    unsigned int ndetParams;
    in.read(reinterpret_cast<char *>(&ndetParams), sizeof(unsigned int));
    m_detParamsGPU.resize(ndetParams);
    in.read(reinterpret_cast<char *>(m_detParamsGPU.data()), ndetParams * sizeof(pixelCPEforGPU::DetParams));
    in.read(reinterpret_cast<char *>(&m_averageGeometry), sizeof(pixelCPEforGPU::AverageGeometry));
    in.read(reinterpret_cast<char *>(&m_layerGeometry), sizeof(pixelCPEforGPU::LayerGeometry));
  }

  cpuData_ = {
      &m_commonParamsGPU,
      m_detParamsGPU.data(),
      &m_layerGeometry,
      &m_averageGeometry,
  };
}

File: src/julia-serial/CondFormats/PixelCPEFast.jl
using .PixelGPU_h
using .Geometry_TrackerGeometryBuilder_phase1PixelTopology_h.phase1PixelTopology

struct PixelCPEFast
    m_detParamsGPU::Vector{DetParams}
    m_commonParamsGPU::CommonParams
    m_layerGeometry::LayerGeometry
    m_averageGeometry::AverageGeometry
    cpuData_::ParamsOnGPU
end



# Define the getCPUProduct method
function getCPUProduct(pixelCPE::PixelCPEFast)::ParamsOnGPU
    return pixelCPE.cpuData_
end

